### PR TITLE
fix: all written artifacts should be saved and garbage collected

### LINKED ISSUE #13583: ArtifactGC fails if output artifact with non-existent file is configured with `optional: true`
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened? What did you expect to happen?

What happened?:
   - When configuring Artifact Garbage Collection in a workflow, if the workflow reaches it's deadline before it manages to create a file which is to be used as an output artifact and that artifact is configured with `optional: true` parameter, then the ArtGC pod which runs after the steps are terminated fails with the message: `Error: You need to configure artifact storage. More information on how to do this can be found in the docs: https://argoproj.github.io/argo-workflows/configure-artifact-repository/`
   -  If you remove the `optional: true` config from the offending artifact and re-run, then no artifact GC pod is created when the workflow terminates from deadline exceeded.
   - If you remove the offending artifact item entirely and re-run, no such `Error: You need to configure artifact storage` messages appear and the other artifacts are cleaned as expected when timeout occurs.

What did I expect to happen:
  - I expected the existing artifacts to be cleaned up and the non-existent artifact to be ignored in the ArtGC pod. The ArtGC pod should finish with a success

### Version(s)

v3.4.10, v3.5.10, latest(sha256:4f03ff7ecaef4061dddd2c08f80de4d766b253aa3a57a87e69dd3a797bb42b1e)

### Paste a minimal workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: test-artgc-
  namespace: argo
spec:
  entrypoint: main
  activeDeadlineSeconds: 20

  volumes:
    - name: emptydir
      emptyDir: {}

  templates:
    - name: main
      dag:
        tasks:
          - name: prepare-data
            template: prepare-data
            arguments:
              parameters: 
              - name: index
                value: "{{item}}"
            withParam: '[1, 2, 3]'
          
    - name: prepare-data
      inputs:
        parameters:
          - name: index
          - name: image
            value: "alpine:latest"
      outputs:
        artifacts:
          # Comment out the 'late-artifact' item and re-run to get a
          # successful artifact garbage collection pod execution
          - name: late-artifact-{{inputs.parameters.index}}
            path: "/empty-dir/late-artifact.txt"
            artifactGC:
              strategy: OnWorkflowCompletion
            optional: true  # removing this will result in no ArtGC pod being run
          - name: abcartifact-{{inputs.parameters.index}}
            path: "/empty-dir/abcartifact.txt"
            artifactGC:
              strategy: OnWorkflowCompletion
            optional: true
          - name: 123artifact-{{inputs.parameters.index}}
            path: "/empty-dir/123artifact.txt"
            artifactGC:
              strategy: OnWorkflowCompletion
            optional: true
      container:
        image: "{{inputs.parameters.image}}"
        volumeMounts:
          - name: emptydir
            mountPath: /empty-dir
        workingDir: /empty-dir
        command: ["sh", "-c"]
        args:
          - |

            set -x
            
            echo Hi > abcartifact.txt
            echo Hi > 123artifact.txt

            # cause timeout
            sleep 20

            echo Hi > late-artifact.txt
```

### Logs from the workflow controller

```text
$  kubectl logs -n argo deploy/workflow-controller | grep test-artgc-pv7h2

time="2024-09-09T15:50:26.593Z" level=info msg="Processing workflow" Phase=Failed ResourceVersion=209475215 namespace=argo workflow=test-artgc-pv7h2
time="2024-09-09T15:50:26.594Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=0 workflow=test-artgc-pv7h2
time="2024-09-09T15:50:36.472Z" level=info msg="Processing workflow" Phase=Failed ResourceVersion=209475215 namespace=argo workflow=test-artgc-pv7h2
time="2024-09-09T15:50:36.472Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=0 workflow=test-artgc-pv7h2
```

### Logs from in your workflow's wait container

```text
$  kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=test-artgc-pv7h2,workflow.argoproj.io/phase!=Succeeded

error: container wait is not valid for pod test-artgc-pv7h2-artgc-wfcomp-2166136261
```

### PR DESCRIPTION
**Pull request by juliev0**
Fixes #13583 

### Motivation

Two issues:
1. Optional artifacts that were not written were still being passed back through the `WorkflowTaskResult`, which was causing Artifact GC to try to delete them (and fail to do so)
2. If any non-optional artifact did fail to get saved because it wasn't present, other artifacts would not get saved

### Modifications

To address # 1 above, only write to `WorkflowTaskResult` the artifacts which have been written, whether or not they're marked "Optional". This is what Artifact GC will attempt to delete.

To address # 2 above, don't prematurely return if an artifact fails to be saved.

### Verification

e2e test for Artifact GC now includes tests for these cases. 
To test # 1 above, see [artgc-optional-artifact-not-written.yaml](https://github.com/argoproj/argo-workflows/pull/13678/files#diff-425d1aaa0cb7f000f0b87bfcdbe2389784b658074e26aca417fc238560095ed2), which includes an Optional artifact that isn't present, which we confirm does not fail ArtifactGC. We also confirm that the Workflow succeeds in this case.

To test # 2 above, see [artgc-non-optional-artifact-not-written.yaml ](https://github.com/argoproj/argo-workflows/pull/13678/files#diff-683e9c31fb266e08c908dd874acef8d9af11a0594694d52a228260eeee53b888), which includes a non-Optional artifact that isn't present, plus another artifact which is present and we confirm does get saved. We also confirm that the Workflow succeeds in this case, since the Artifact that wasn't present was non-optional.

### CODE REVIEW

File: test/e2e/artifacts_test.go [hunk 1]
```diff
@@ -307,11 +307,11 @@ func (s *ArtifactsSuite) TestArtifactGC() {
 			workflowShouldSucceed:        true,
 			expectedGCPodsOnWFCompletion: 2,
 			expectedArtifacts: []artifactState{
-				artifactState{s3Location{bucketName: "my-bucket-2", specifiedKey: "first-on-completion-1"}, true, false},
-				artifactState{s3Location{bucketName: "my-bucket-3", specifiedKey: "first-on-completion-2"}, true, false},
-				artifactState{s3Location{bucketName: "my-bucket-3", specifiedKey: "first-no-deletion"}, false, false},
-				artifactState{s3Location{bucketName: "my-bucket-3", specifiedKey: "second-on-deletion"}, false, true},
-				artifactState{s3Location{bucketName: "my-bucket-2", specifiedKey: "second-on-completion"}, true, false},
+				{s3Location{bucketName: "my-bucket-2", specifiedKey: "first-on-completion-1"}, true, false},
```
Comment (juliev0) [inline test/e2e/artifacts_test.go:310]: removing redundant type specification as recommended by linter

**Comment by juliev0** [comment]
Hey @agilgur5 - I know you mentioned saving artifacts in parallel would be a good improvement. My thought was that with what I did it's clearly just a "fix" (which could theoretically be patched), and that saving in parallel would be more of a feature. I was also paranoid there could be some new risk of doing it in parallel, that some artifact type may not be able to handle the concurrent saves, although maybe that's not likely. If you feel this is silly, I can revise.

**Review by isubasinghe — APPROVED** [review]
LGTM

File: workflow/executor/executor.go [hunk 2]
```diff
@@ -306,39 +306,46 @@ func (we *WorkflowExecutor) SaveArtifacts(ctx context.Context) (wfv1.Artifacts,
...
+	var firstArtifactError error
 	for _, art := range we.Template.Outputs.Artifacts {
-		err := we.saveArtifact(ctx, common.MainContainerName, &art)
-		if err != nil {
-			return artifacts, err
+		saved, err := we.saveArtifact(ctx, common.MainContainerName, &art)
+
+		if err != nil && firstArtifactError == nil {
+			firstArtifactError = err
+		}
```
[the remark this answers is not in the document]
Comment (juliev0) [reply workflow/executor/executor.go]: sounds good. My original thought was that usually if one fails, the others are failing for the same reason. But doing the aggregate error is more complete and doesn't mislead them into thinking maybe only just one artifact failed. I have revised.

**Pull request merged**